### PR TITLE
Implement discv5 topic registration

### PR DESCRIPTION
**Pull request by gsalgado**
This basically adds two new APIs that allow us to ask a remote node for
a discv5 ticket (DiscoveryProtocol.get_ticket()) and then use that to
register ourselves under a given topic (DiscoveryProtocol.send_topic_register())

### CODE REVIEW

File: p2p/discovery.py [hunk 1]
```diff
@@ -765,6 +791,45 @@ def process(received_echo: Hash32, response: List[kademlia.Node]) -> None:
 
         return tuple(n for n in nodes if n != self.this_node)
 
+    async def get_ticket(self, node: kademlia.Node, topics: List[bytes]) -> 'Ticket':
+        token = self.send_ping_v5(node, topics)
+        try:
+            got_pong, raw_pong, wait_periods = await self.wait_pong_v5(node, token)
+        except AlreadyWaitingDiscoveryResponse:
+            self.logger.debug("failed to get ticket from %s, already waiting for pong", node)
+            return None
```
Comment (pipermerriam) [inline p2p/discovery.py]: I'm generally :+1: on APIs that require return value checking as they are prone to errors when someone fails to actually check.

File: p2p/discovery.py [hunk 2]
```diff
@@ -765,6 +791,45 @@ def process(received_echo: Hash32, response: List[kademlia.Node]) -> None:
 
         return tuple(n for n in nodes if n != self.this_node)
 
+    async def get_ticket(self, node: kademlia.Node, topics: List[bytes]) -> 'Ticket':
```
Comment (pipermerriam) [inline p2p/discovery.py]: See my comment below about return value checking but.... the signature for this function says it returns a `Ticket` when it should really be `Optional[Ticket]` right?

File: tests/p2p/test_discovery.py [hunk 3]
```diff
@@ -431,6 +450,14 @@ def get_discovery_protocol(seed=b"seed", address=None):
     return discovery.DiscoveryProtocol(privkey, address, [], CancelToken("discovery-test"))
 
 
+async def get_listening_discovery_protocol(event_loop):
+    addr = kademlia.Address('127.0.0.1', random.randint(1024, 9999))
```
Comment (pipermerriam) [inline tests/p2p/test_discovery.py]: Just noting that I looked (briefly) and it appears that there are no obvious ways to get a random open UDP port

File: p2p/discovery.py (hunk 1 above)
Comment (gsalgado) [reply p2p/discovery.py]: I guess you mean :-1:?

**Comment by gsalgado** [comment]
I agree that `DiscoveryProtocol` may be getting unnecessarily complex; I'll have a look into ways to simplify it

File: p2p/discovery.py (hunk 2 above)
Comment (gsalgado) [reply p2p/discovery.py]: That's right

File: p2p/discovery.py (hunk 1 above)
Comment (gsalgado) [reply p2p/discovery.py]: Anyway, I've changed it to raise an exception instead

Comment (pipermerriam) [reply p2p/discovery.py]: :joy_cat: yup, :-1:

**Pull request merged**